hook-common/src/pgqueue.rs

```diff
@@ -441,42 +385,127 @@ RETURNING
                 error,
             })?;
 
-        Ok(retryable_job)
+        Ok(retried_job)
     }
 }
 
 /// A Job that has failed but can still be enqueued into a PgQueue to be retried at a later point.
 /// The time until retry will depend on the PgQueue's RetryPolicy.
-pub struct RetryableJob<J> {
+pub struct RetryableJob {
     /// A unique id identifying a job.
-    id: i64,
+    pub id: i64,
     /// A number corresponding to the current job attempt.
     pub attempt: i32,
-    /// Any JSON-serializable value to be stored as an error.
-    pub error: sqlx::types::Json<J>,
     /// A unique id identifying a job queue.
-    pub queue: String,
+    queue: String,
+    /// An optional separate queue where to enqueue this job when retrying.
+    retry_queue: Option<String>,
+}
+
+impl RetryableJob {
+    /// Set the queue for a `RetryableJob`.
+    /// If not set, `Job` will be retried to its original queue on calling `retry`.
+    fn queue(mut self, queue: &str) -> Self {
+        self.retry_queue = Some(queue.to_owned());
+        self
+    }
+
+    /// Return the queue that a `Job` is to be retried into.
+    fn retry_queue(&self) -> &str {
+        self.retry_queue.as_ref().unwrap_or(&self.queue)
+    }
+
+    /// Consume `Job` to retry it.
+    /// A `RetriedJob` cannot be used further; it is returned for reporting or inspection.
+    ///
+    /// # Arguments
+    ///
+    /// * `error`: Any JSON-serializable value to be stored as an error.
+    /// * `retry_interval`: The duration until the `Job` is to be retried again. Used to set `scheduled_at`.
+    /// * `table`: The table where this job will be marked as completed.
+    /// * `executor`: Any sqlx::Executor that can execute the UPDATE query required to mark this `Job` as completed.
+    async fn retry<'c, S, E>(
+        self,
+        error: S,
+        retry_interval: time::Duration,
+        table: &str,
+        executor: E,
+    ) -> Result<RetriedJob, sqlx::Error>
+    where
+        S: serde::Serialize + std::marker::Sync + std::marker::Send,
+        E: sqlx::Executor<'c, Database = sqlx::Postgres>,
+    {
+        let json_error = sqlx::types::Json(error);
+        let base_query = format!(
+            r#"
+UPDATE
+    "{0}"
+SET
+    finished_at = NOW(),
```

Comment:
Should this be `attempted_at`? I feel like we discussed some confusion around these before, so I guess they're still confusing to me. :P

Reply:
Ahhh, https://github.com/PostHog/rusty-hook/pull/7#discussion_r1426858428

It's more like `last_attempt_finished_at`, I had to refresh my memory.